components/script/dom/servoparser/async_html.rs

```diff
@@ -0,0 +1,561 @@
+        let document = Root::from_ref(&**self.nodes.borrow().get(&0).expect("document node should exist!"));
+        let document = document.downcast::<Document>().expect("Document node should be downcasted!");
+        match op {
+            ParseOperation::GetTemplateContents(target, contents) => {
+                let target = Root::from_ref(&**self.nodes.borrow().get(&target).unwrap());
+                let template = target.downcast::<HTMLTemplateElement>().expect(
+                    "Tried to extract contents from non-template element while parsing");
+                assert!(self.nodes.borrow_mut().insert(contents, JS::from_ref(template.Content().upcast())).is_none());
+            }
+            ParseOperation::CreateElement(id, name, attrs) => {
+                let elem = Element::create(name, &*self.document,
+                                           ElementCreator::ParserCreated(self.current_line));
+                for attr in attrs {
+                    elem.set_attribute_from_parser(attr.name, DOMString::from(String::from(attr.value)), None);
+                }
+
+                self.nodes.borrow_mut().insert(id, JS::from_ref(elem.upcast()));
+            }
+            ParseOperation::CreateComment(text, id) => {
+                let comment = Comment::new(DOMString::from(text), document);
+                assert!(self.nodes.borrow_mut().insert(id, JS::from_ref(&comment.upcast())).is_none());
+            }
+            ParseOperation::Insert(parent, reference_child, new_node) => {
+                let nodes = self.nodes.borrow();
+                let parent = nodes.get(&parent).unwrap();
+                let reference_child = reference_child.map(|n| nodes.get(&n).unwrap().deref());
+
+                let node = match new_node {
+                    ParserNodeOrText::Node(ref id) => {
+                        let deref_node = nodes.get(id).unwrap().deref();
+                        NodeOrText::AppendNode(JS::from_ref(deref_node))
+                    },
+                    ParserNodeOrText::Text(ref string) => {
+                        match Tendril::from_str(string) {
+                            Ok(tendril) => NodeOrText::AppendText(tendril),
+                            Err(_) => panic!("Error while converting from string to tendril!"),
+                        }
+                    }
+                };
+
+                super::insert(parent, reference_child, node);
+            }
+            ParseOperation::AppendDoctypeToDocument(name, public_id, system_id) => {
+                let doctype = DocumentType::new(
+                    DOMString::from(name), Some(DOMString::from(public_id)),
+                    Some(DOMString::from(system_id)), document);
+
+                document.upcast::<Node>().AppendChild(doctype.upcast()).expect("Appending failed");
+            }
+            ParseOperation::AddAttrsIfMissing(target_id, attrs) => {
+                match self.nodes.borrow().get(&target_id) {
```

Comment:
Make a private `node` method on `self` that returns a node from the `nodes` map or panics if there are none.